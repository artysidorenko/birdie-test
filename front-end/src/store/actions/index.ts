import { ActionTypes, actions } from '../types';

export function getEvents(): ActionTypes {
  return {
    type: actions.GET_EVENTS
  };
}
export function getFluidIntake(): ActionTypes {
  return {
    type: actions.GET_FLUID_INTAKE
  };
}
export function getFoodIntake(): ActionTypes {
  return {
    type: actions.GET_FOOD_INTAKE
  };
}
export function getMedication(): ActionTypes {
  return {
    type: actions.GET_MEDICATION
  };
}
export function getMoods(): ActionTypes {
  return {
    type: actions.GET_MOODS
  };
}
export function getObservations(): ActionTypes {
  return {
    type: actions.GET_OBSERVATIONS
  };
}
export function getPadCondition(): ActionTypes {
  return {
    type: actions.GET_PAD_CONDITION
  };
}
export function getTasks(): ActionTypes {
  return {
    type: actions.GET_TASKS
  };
}