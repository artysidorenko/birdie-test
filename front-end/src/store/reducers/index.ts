import { combineReducers, Reducer } from 'redux';
import { actions, DataState } from '../types';
import { initialState } from './initialState';

export type RootState = Readonly<{
  readonly data: DataState
}>;

export const dataReducer: Reducer<DataState> = (state = initialState, action) => {
  switch (action.type) {
    case actions.SET_ID:
      return {
        ...state,
        id: action.id
      };
    case actions.GET_EVENTS_RESPONSE:
      return {
        ...state,
        events: {
          ...state.events,
          status: action.status,
          ...(action.status === 'failure' && {error: action.response}),
          ...(action.status === 'success' && {events: action.response})
        }
      };
    case actions.GET_FLUID_INTAKE_RESPONSE:
      return {
        ...state,
        fluid_intake: {
          ...state.fluid_intake,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { fluid_intake: action.response })
        }
      };
    case actions.GET_FOOD_INTAKE_RESPONSE:
      return {
        ...state,
        food_intake: {
          ...state.food_intake,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { food_intake: action.response })
        }
      };
    case actions.GET_MEDICATION_RESPONSE:
      return {
        ...state,
        medication: {
          ...state.medication,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { medication: action.response })
        }
      };
    case actions.GET_MOODS_RESPONSE:
      return {
        ...state,
        moods: {
          ...state.moods,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { moods: action.response })
        }
      };
    case actions.GET_OBSERVATIONS_RESPONSE:
      return {
        ...state,
        observations: {
          ...state.observations,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { observations: action.response })
        }
      };
    case actions.GET_PAD_CONDITION_RESPONSE:
      return {
        ...state,
        pad_condition: {
          ...state.pad_condition,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { pad_condition: action.response })
        }
      };
    case actions.GET_TASKS_RESPONSE:
      return {
        ...state,
        tasks: {
          ...state.tasks,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { tasks: action.response })
        }
      };
    default:
      return state;
  }
};

export const rootReducer = combineReducers<RootState>({
  data: dataReducer
});