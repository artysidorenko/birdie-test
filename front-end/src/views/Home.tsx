import * as React from 'react';

const Dashboard = () => (
  <>
    Welcome to the Care Dashboard.
    Here you can view information about the care being received by the receipient,
    including updates on their mood and health.
  </>
);

export default Dashboard;
