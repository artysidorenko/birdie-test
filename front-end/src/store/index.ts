import { applyMiddleware, compose, createStore, GenericStoreEnhancer } from 'redux';
import createSagaMiddleware from 'redux-saga';
import * as _history from 'history';
import { rootReducer, RootState } from '@App/store/reducers';
import initSaga from '@App/store/sagas';
import { DataState } from './types';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION__: () => undefined;
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__: (arg: GenericStoreEnhancer) => undefined;
  }
}

const sagaMiddleware = createSagaMiddleware();
export const history = _history.createBrowserHistory();

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);

sagaMiddleware.run(initSaga);

export default store;

export const mapStateToProps = (state: RootState, ownProps: object): DataState => ({
  events: state.data.events,
  fluid_intake: state.data.fluid_intake,
  food_intake: state.data.food_intake,
  observations: state.data.observations,
  pad_condition: state.data.pad_condition,
  moods: state.data.moods,
  medication: state.data.medication,
  tasks: state.data.tasks
});