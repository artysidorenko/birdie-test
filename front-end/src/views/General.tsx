import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import styled from 'styled-components';

import { RootState } from '@App/store/reducers';
import { mapStateToProps } from '../store';
import { DataState, status as fetchStatus } from '../store/types';
import { getEvents, getMoods } from '../store/actions';

import VisitsBar from '@App/components/graphs/VisitsBar';
import CarerPie from '@App/components/graphs/CarerPie';
import MoodsBar from '@App/components/graphs/MoodsBar';

interface ActionProps {
  getEvents: () => void;
  getMoods: () => void;
}

type AppProps = DataState & ActionProps;

const cardFormat = `
  box-shadow: 0 0.46875rem 2.1875rem rgba(90,97,105,.1),
    0 0.9375rem 1.40625rem rgba(90,97,105,.1),
    0 0.25rem 0.53125rem rgba(90,97,105,.12),
    0 0.125rem 0.1875rem rgba(90,97,105,.1);

  padding: 10px;
  margin-bottom: 20px;
  `;

const StyledWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;

  @media screen and (max-width: 950px) {
    flex-direction: column;
    align-items: center;
  }
`;

const StyledContainer = styled.div`
  position: relative;
  height: 40vh;
  width: 82.5vw;
  ${cardFormat}
`;
const StyledContainerSmall = styled.div`
  position: relative;
  height: 40vh;
  width: 40vw;
  margin-left: 10px;
  margin-right: 10px;
  ${cardFormat}

  @media screen and (max-width: 950px) {
    width: 80vw;
    margin-left: 0px;
    margin-right: 0px;
  }
`;
const Row = styled.div`
  position: relative;
  display: flex;
  justify-content: space-evenly;

  @media screen and (max-width: 950px) {
    flex-direction: column;
    align-items: center;
  }
`;

class General extends React.Component<AppProps> {
  displayName: '';

  componentDidMount() {
    this.props.getEvents();
    this.props.getMoods();
  }

  render() {
    const { status: eventStatus, events } = this.props.events;
    const { status: moodStatus, moods } = this.props.moods;

    return (
      <StyledWrapper>
        {eventStatus === fetchStatus.SUCCESS &&
        moodStatus === fetchStatus.SUCCESS ? (
          <React.Fragment>
            <Row>
              <StyledContainerSmall id="mood">
                <MoodsBar data={moods} />
              </StyledContainerSmall>
              <StyledContainerSmall id="carers">
                <CarerPie data={events} />
              </StyledContainerSmall>
            </Row>
            <StyledContainer id="visits">
              <VisitsBar data={events} />
            </StyledContainer>
          </React.Fragment>
        ) : (
          <span>Loading...</span>
        )}
      </StyledWrapper>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<RootState>) => ({
  getEvents: () => dispatch(getEvents()),
  getMoods: () => dispatch(getMoods())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(General);