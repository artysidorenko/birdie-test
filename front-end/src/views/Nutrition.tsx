import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import styled from 'styled-components';

import { RootState } from '@App/store/reducers';
import { mapStateToProps } from '../store';
import { DataState, status as fetchStatus } from '../store/types';
import { getFluidIntake, getFoodIntake } from '../store/actions';

import FluidGraph from '@App/components/graphs/FluidGraph';
import FoodGraph from '@App/components/graphs/FoodGraph';

interface ActionProps {
  getFluidIntake: () => void;
  getFoodIntake: () => void;
}

type AppProps = DataState & ActionProps;

const StyledContainer = styled.div`
  position: relative;
  height: 50vh;
`;

class Nutrition extends React.Component<AppProps> {
  displayName: 'LineExample';

  componentDidMount() {
    this.props.getFluidIntake();
    this.props.getFoodIntake();
  }

  render() {
    const { status: fluidStatus, fluid_intake } = this.props.fluid_intake;
    const { status: foodStatus, food_intake } = this.props.food_intake;

    return (
      <div>
        {fluidStatus === fetchStatus.SUCCESS ? (
          <StyledContainer>
            <FluidGraph data={fluid_intake} />
          </StyledContainer>
        ) : (
          <span>Loading...</span>
        )}
        {foodStatus === fetchStatus.SUCCESS ? (
          <StyledContainer>
            <FoodGraph data={food_intake} />
          </StyledContainer>
        ) : (
          <span>Loading...</span>
        )}
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<RootState>) => ({
  getFluidIntake: () => dispatch(getFluidIntake()),
  getFoodIntake: () => dispatch(getFoodIntake())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Nutrition);
