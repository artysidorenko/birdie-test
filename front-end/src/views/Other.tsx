import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import styled from 'styled-components';

import { RootState } from '@App/store/reducers';
import { mapStateToProps } from '../store';
import { DataState, status as fetchStatus } from '../store/types';
import { getMedication, getPadCondition, getTasks } from '../store/actions';

import MixedBar from '@App/components/graphs/MixedBar';
import TaskGraph from '@App/components/graphs/TaskGraph';

interface ActionProps {
  getMedication: () => void;
  getPadCondition: () => void;
  getTasks: () => void;
}

type AppProps = DataState & ActionProps;

const cardFormat = `
  box-shadow: 0 0.46875rem 2.1875rem rgba(90,97,105,.1),
    0 0.9375rem 1.40625rem rgba(90,97,105,.1),
    0 0.25rem 0.53125rem rgba(90,97,105,.12),
    0 0.125rem 0.1875rem rgba(90,97,105,.1);

  padding: 10px;
  margin-bottom: 20px;
  `;

const StyledContainer = styled.div`
  position: relative;
  height: 70vh;
  width: 40vw;
  ${cardFormat}

  @media screen and (max-width: 950px) {
    position: relative;
    height: 50vh;
    width: 80vw;
    margin-left: 20px;
    margin-right: 20px;
  }
`;

const Row = styled.div`
  position: relative;
  display: flex;
  justify-content: space-evenly;

  @media screen and (max-width: 950px) {
    flex-direction: column;
  }
`;

class Other extends React.Component<AppProps> {
  displayName: 'LineExample';

  componentDidMount() {
    this.props.getMedication();
    this.props.getPadCondition();
    this.props.getTasks();
  }

  render() {
    const { status: eventStatus, medication } = this.props.medication;
    const { status: padStatus, pad_condition } = this.props.pad_condition;
    const { status: taskStatus, tasks } = this.props.tasks;

    return (
      <Row>
        {eventStatus === fetchStatus.SUCCESS &&
        padStatus === fetchStatus.SUCCESS &&
        taskStatus === fetchStatus.SUCCESS ? (
          <React.Fragment>
            <StyledContainer id="mixed">
              <MixedBar medData={medication} padData={pad_condition} />
            </StyledContainer>
            <StyledContainer id="tasks">
              <TaskGraph data={tasks} />
            </StyledContainer>
          </React.Fragment>
        ) : (
          <span>Loading...</span>
        )}
      </Row>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<RootState>) => ({
  getMedication: () => dispatch(getMedication()),
  getPadCondition: () => dispatch(getPadCondition()),
  getTasks: () => dispatch(getTasks())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Other);
