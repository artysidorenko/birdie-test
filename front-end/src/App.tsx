import * as React from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import { Route } from 'react-router-dom';

import Menu from '@App/components/Menu';

import Home from './views/Home';
import General from './views/General';
import Nutrition from './views/Nutrition';
import Other from './views/Other';

interface ActionProps {
}

type AppProps = ActionProps;

interface AppState {

}

const GlobalStyle = createGlobalStyle`
  body {
    height: 100vh;
    background-color: #F9F9F9;
    > div {
      height: 100%;
    }
  }
`;

const AppContainer = styled.div`
  width: 100%;
  height: 100%;
`;

class App extends React.Component<AppProps, AppState> {
  public constructor(props: AppProps) {
    super(props);
  }
  componentDidMount() {
  }

  public render() {
    /* tslint:disable:no-console */
    // console.log(this.props.events.status, this.props.events.events);
    return (
      <>
        <GlobalStyle />
        <AppContainer>
          <Menu />
          <Route exact={true} path="/" component={Home} />
          <Route path="/general" component={General} />
          <Route path="/nutrition" component={Nutrition} />
          <Route path="/other" component={Other} />
        </AppContainer>
      </>
    );
  }
}

export default App;