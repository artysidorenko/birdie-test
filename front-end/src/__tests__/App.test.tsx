import * as React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import * as chai from 'chai';
import { expect as _expect } from 'chai';

import App from '../App';
import store from '@App/store';
import General from '@App/views/General';
import Home from '@App/views/Home';
import Nutrition from '@App/views/Nutrition';
import Other from '@App/views/Other';

import matcher from 'chai-react-element';
chai.use(matcher);


/* Normally part of src/setupTests.js but ... */
import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
/* ... file was not being picked up by CRA so added manually */

/* tslint:disable:no-unused-expression */

describe('Application', () => {

  it('renders without crashing (shallow)', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders without crashing (mount)', () => {
    const wrapper = mount(<Provider store={store}><Router><App /></Router></Provider>);
    _expect(wrapper).to.exist;
  });

  it('calls componentDidMount', () => {
    const mockHook = jest.spyOn(App.prototype, 'componentDidMount');
    mockHook.mockImplementation(() => { });
    shallow(<App />);
    expect(mockHook.mock.calls.length).toBe(1);
  });

  // it('renders 4 views on mounting', () => {
  //   const wrapper = mount(<Provider store={store}><Router><App /></Router></Provider>);
  //   _expect(wrapper.containsMatchingElement(<General/>)).to.be.true;
  // });

});
