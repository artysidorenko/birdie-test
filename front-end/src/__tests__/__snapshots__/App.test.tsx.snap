// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders without crashing (shallow) 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styled.div>
    <MenuDesktop />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      path="/general"
    />
    <Route
      component={[Function]}
      path="/nutrition"
    />
    <Route
      component={[Function]}
      path="/other"
    />
  </styled.div>
</Fragment>
`;
